gpt4all-bindings/typescript/src/gpt4all.js

```diff
@@ -0,0 +1,112 @@
+/// This file implements the gpt4all.d.ts file endings.
+/// Written in commonjs to support both ESM and CJS projects.
+
+const { LLModel } = require('bindings')('../build/Release/gpt4allts');
+const { createWriteStream, existsSync } = require('fs');
+const { join } = require('path');
+const { performance } = require('node:perf_hooks');
+
+
+
+// readChunks() reads from the provided reader and yields the results into an async iterable
+// https://css-tricks.com/web-streams-everywhere-and-fetch-for-node-js/
+function readChunks(reader) {
+    return {
+        async* [Symbol.asyncIterator]() {
+            let readResult = await reader.read();
+            while (!readResult.done) {
+                yield readResult.value;
+                readResult = await reader.read();
+            }
+        },
+    };
+}
+
+exports.LLModel = LLModel;
+
+
+exports.download = function (
+    name,
+    options = { debug: false, location: process.cwd(), link: undefined }
+) {
+    const abortController = new AbortController();
+    const signal = abortController.signal;
+
+    const pathToModel = join(options.location, name);
+    if(existsSync(pathToModel)) {
+        throw Error("Path to model already exists");
+    }
+
+    //wrapper function to get the readable stream from request
+    const fetcher = (name) => fetch(options.link ?? `https://gpt4all.io/models/${name}`, {
+        signal,
+    })
+    .then(res => {
+         if(!res.ok) {
+            throw Error("Could not find "+ name + " from " + `https://gpt4all.io/models/` )
+         }
+         return res.body.getReader()
+    })
+    
+    //a promise that executes and writes to a stream. Resolves when done writing.
+    const res = new Promise((resolve, reject) => {
+        fetcher(name)
+            //Resolves an array of a reader and writestream.
+            .then(reader => [reader, createWriteStream(pathToModel)])
+            .then( 
+            async ([readable, wstream]) => {
+                console.log('(CLI might hang) Downloading @ ', pathToModel);
+                let perf;
+                if(options.debug) {
+                   perf = performance.now(); 
+                }
+                for await (const chunk of readChunks(readable)) {
+                    wstream.write(chunk);
+                }
+                if(options.debug) {
+                    console.log("Time taken: ", (performance.now()-perf).toFixed(2), " ms"); 
+                }
+                resolve();
+            }
+        ).catch(reject);
+    });
+    
+    return {
+        cancel : () => abortController.abort(),
+        promise: () => res
+    }
+}
+
+
+//https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates
+exports.prompt = function prompt(strings, ...keys) {
```

Comment:
Is there a rationale about why you went with a tagged template?

As a user, this design feels quite complicated to use or even understand, and also quite different from the Python one ( see [this file](https://github.com/nomic-ai/gpt4all/blob/73db20ba85fbbdc66a56e2619394c0eea40dc72b/gpt4all-bindings/python/gpt4all/gpt4all.py#L188))

Reply:
Hey, it was a spur of the moment kind of thing that I thought would be creative. As the bindings are still in alpha, this is changeable :), however, what part of the api seems complex to you?

Reply:
```
prompt`${"header"} ${"prompt"}`
``` is very unusual since it differs so much from a regular template literal (which would just print "header prompt" on the given example). I do not expect a tagged template to return a function.
I would go with a function instead of a tagged template for more clarity.

But to step back even more, why is the API not the same as the Python one?

`createCompletion(ll, [{role: "user", content:"hello"])`

Reply:
Thanks for the feedback. I could always implement it. It's better to stay consistent with python bindings as you suggested.